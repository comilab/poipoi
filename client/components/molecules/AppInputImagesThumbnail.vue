<template lang="pug">
.relative.border.rounded.bg-white.cursor-move.p-1
  AppSquare
    img.object-contain(
      :src="thumbnail",
      data-testid="thumbnail"
    )
    .bg-white.opacity-50.absolute.inset-0(
      v-if="image.active || image.success",
      data-testid="progress"
    )
      font-awesome-icon.position-center.fa-2x.text-black(
        :icon="progressIcon",
        :pulse="image.active"
      )
  .absolute.right-0.bottom-0.bg-white.rounded.text-center.text-sm.w-8.h-6 {{ num }}
  AppButton.remove-button.absolute.shadow.w-8.h-8(
    color="blue",
    size="sm",
    :icon="faTimes",
    :disabled="image.active || image.success",
    pill,
    @click.stop="$emit('remove')",
    data-testid="remove-button"
  )
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import { PropType } from '@vue/composition-api'
import { faTimes, faSpinner, faCheck } from '@fortawesome/free-solid-svg-icons'

import Image from '~/models/Image'
import AppSquare from '~/components/atoms/AppSquare.vue'
import AppButton from '~/components/atoms/AppButton.vue'

export default defineComponent({
  components: {
    AppSquare,
    AppButton
  },
  props: {
    image: {
      type: Object as PropType<VUFile|Image>,
      required: true
    },
    num: {
      type: Number,
      required: true
    }
  },
  setup (props) {
    const URL = window.URL || window.webkitURL

    const thumbnail = computed(() => {
      if (props.image instanceof Image) {
        return props.image.paths.small
      } else if (URL && URL.createObjectURL) {
        return URL.createObjectURL(props.image.file)
      }
      return null
    })

    const progressIcon = computed(() => {
      if (!(props.image instanceof Image)) {
        if (props.image.active) {
          return faSpinner
        }
        if (props.image.success) {
          return faCheck
        }
      }
      return null
    })

    return {
      thumbnail,
      faTimes,
      progressIcon
    }
  }
})
</script>

<style lang="postcss" scoped>
.remove-button {
  top: -1rem;
  right: -1rem;
}
</style>
