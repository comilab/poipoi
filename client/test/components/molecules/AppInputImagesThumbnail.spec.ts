import { fireEvent, render } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'
import { faSpinner, faCheck } from '@fortawesome/free-solid-svg-icons'

import AppInputImagesThumbnail from '~/components/molecules/AppInputImagesThumbnail.vue'
import imageFactory from '~/test/factories/image'

describe('components/molecules/AppInputImagesThumbnail', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        image: {
          file: {},
          active: false,
          success: false
        },
        num: 1
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    const { container } = render(AppInputImagesThumbnail, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('image', () => {
      describe('Imageモデルの場合', () => {
        test('img要素のsrc属性にpaths.smallをセット', () => {
          options.props.image = imageFactory.build()
          const { getByTestId } = render(AppInputImagesThumbnail, options)

          expect(getByTestId('thumbnail')).toHaveAttribute('src', options.props.image.paths.small)
        })
      })

      describe('Imageモデルでない場合', () => {
        let fontAwesomeIconProps: any

        beforeEach(() => {
          options.stubs.FontAwesomeIcon = defineComponent({
            props: ['icon'],
            setup (props) {
              fontAwesomeIconProps = props
            },
            template: '<span />'
          })
        })

        test('activeがtrueなら進捗アイコンとしてfaSpinnerを表示する', async () => {
          const { queryByTestId, updateProps } = render(AppInputImagesThumbnail, options)
          expect(queryByTestId('progress')).not.toBeInTheDocument()

          await updateProps({
            image: {
              file: {},
              active: true,
              success: false
            }
          })
          expect(queryByTestId('progress')).toBeInTheDocument()
          expect(fontAwesomeIconProps.icon).toBe(faSpinner)
        })

        test('successがtrueなら進捗アイコンとしてfaCheckを表示する', async () => {
          const { queryByTestId, updateProps } = render(AppInputImagesThumbnail, options)
          expect(queryByTestId('progress')).not.toBeInTheDocument()

          await updateProps({
            image: {
              file: {},
              active: false,
              success: true
            }
          })
          expect(queryByTestId('progress')).toBeInTheDocument()
          expect(fontAwesomeIconProps.icon).toBe(faCheck)
        })

        test('activeがtrueなら削除ボタンをdisabledにする', async () => {
          const { queryByTestId, updateProps } = render(AppInputImagesThumbnail, options)
          expect(queryByTestId('remove-button')).not.toHaveAttribute('disabled')

          await updateProps({
            image: {
              file: {},
              active: true,
              success: false
            }
          })
          expect(queryByTestId('remove-button')).toHaveAttribute('disabled')
        })

        test('successがtrueなら削除ボタンをdisabledにする', async () => {
          const { queryByTestId, updateProps } = render(AppInputImagesThumbnail, options)
          expect(queryByTestId('remove-button')).not.toHaveAttribute('disabled')

          await updateProps({
            image: {
              file: {},
              active: false,
              success: true
            }
          })
          expect(queryByTestId('remove-button')).toHaveAttribute('disabled')
        })
      })
    })

    describe('num', () => {
      test('numを表示', () => {
        const { getByText } = render(AppInputImagesThumbnail, options)

        getByText('1')
      })
    })
  })

  describe('event', () => {
    describe('remove', () => {
      test('閉じるボタンを押したときに発火', async () => {
        const { getByTestId, emitted } = render(AppInputImagesThumbnail, options)
        await fireEvent.click(getByTestId('remove-button'))

        expect(emitted().remove).toHaveLength(1)
      })
    })
  })
})
