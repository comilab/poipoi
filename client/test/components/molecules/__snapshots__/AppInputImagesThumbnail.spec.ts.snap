// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppInputImagesThumbnail renders correctly 1`] = `
<div>
  <div
    class="relative border rounded bg-white cursor-move p-1"
  >
    <div
      class="is-square relative"
    >
      <div
        class="content absolute inset-0"
      >
        <img
          class="object-contain"
          data-testid="thumbnail"
        />
        <!---->
      </div>
    </div>
    <div
      class="absolute right-0 bottom-0 bg-white rounded text-center text-sm w-8 h-6"
    >
      1
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none remove-button absolute shadow w-8 h-8 bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
      data-testid="remove-button"
      type="button"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      />
    </button>
  </div>
</div>
`;
